Tickets/dm 39591

pcortes80 opened the pull request:
Added: 

- testCreateMultipleTcpIpConnectionsServersAandB.vi,
- testTwoClientsSendCmdAtSameTime.vi,
- testTwoClientsSendEventsAtSameTime.vi,
- testTwoClientsSendTelemetryAtSameTime.vi,
- testTwoServersReportConnStatus.vi,
- testTwoServersReportInternalError.vi,
- testTwoServersReportLogMessages.vi,
- testTwoServersSendCmdStatusToClientsAtSameTime.vi, 
- testTwoServersSendEventsToClientsAtSameTime.vi,
- testTwoServersSendTelemetryAtSameTime.vi, and
- testUserStopsServersAandB.vi

to the `ServerInstancesTest` directory.

teweitsai reviewed (approved):
[Overview]

1. For all the test cases, only the error line of server A will tunnell with a flat structure with waiting time = 2000 ms. I think it is better to let the error line of server B to tunnell to that flat structure as well.


[tests/ServerInstancesTest/disconnectTcpClientA.vi]

1. This is the same as disconnectTcpClientB.vi. Maybe rename this to disconnectTcpClient.vi and remove disconnectTcpClientB.vi.

2. The "label" in block diagram overlaps with the flat structure. Fix this.


[tests/ServerInstancesTest/disconnectTcpClientB.vi]

1. As comment above, remove this vi.


[tests/ServerInstancesTest/readTcpPacketAsString.vi]

1. Please consider to put the timeout=1000 in block diagram to the front panel as an input (+ tunnel). You might want to be able to tune this in the test cases.


[tests/ServerInstancesTest/sendCmdFromTcpClientA.vi]

1. The is the same as sendCmdFromTcpClientB.vi. Maybe rename this to sendCmdFromTcpClient.vi and remove sendCmdFromTcpClientB.vi.

2. Rename the variable of connectionID_A to connectionID.

3. You do not need the shift register of connectionID on the for-loop. The "TCP Write" will not modify that connection ID. You can just use a terminal.

4. Please consider to put the timeout=1000 in block diagram to the front panel as an input (+ tunnel). You might want to be able to tune this in the test cases.

5. If you need to two instances to run at the same time, you can pre-allocate them.


[tests/ServerInstancesTest/sendCmdFromTcpClientB.vi]

1. As comment above, remove this vi.


[tests/ServerInstancesTest/sendEventsFromTcpClient.vi]

1. Please consider to put the timeout=1000 in block diagram to the front panel as an input (+ tunnel). You might want to be able to tune this in the test cases.



[testCase/testCreateMultipleTcpIpConnectionsServersAandB.vi]

1. After the disconnection between the client and server, you would need to wait some time to let the OS to recover the system resource. Otherwise, you might only get the initial connection successfully and all the others failed.

2. Unless there is a good reason, I would prefer to use the shift register for the error line for the for-loop instead of using the tunnel. In addition, I noticed that your called the "Clear Errors.vi" already.


[testCase/testTwoClientsSendCmdAtSameTime.vi]

1. For the vi icon, I think it would be unreadable if you put 4 lines. I think 3 lines are enough. You could use a 13 or 15 inch laptop to verify this.

2. What would happen if you get the empty connection ID (connection fails in timeout) for the command or telemetry server?


[testCase/testTwoClientsSendEventsAtSameTime.vi]

1. For the vi icon, I think it would be unreadable if you put 4 lines. I think 3 lines are enough. You could use a 13 or 15 inch laptop to verify this.

2. There is a logic problem here. You have 5 sec for the connection timeout but only 2 sec for the event case. Do you see the problem?

3. Your event supposes to listern to two EvtRead events but you only put have one in the case structure. You should have two instead. In addition, if you do not care about the received "event", you can judge the result by receiveTimes=2 as an integet (use a shift register to hold the receiveTimes). This is easier to compare than the string array.

4. You need to unregister the events after the event case structure is done and destroy the events to release the system resource.


[testCase/testTwoClientsSendTelemetryAtSameTime.vi]

1. For the vi icon, I think it would be unreadable if you put 4 lines. I think 3 lines are enough. You could use a 13 or 15 inch laptop to verify this.


[testCase/testTwoServersReportConnStatus.vi]

1. For the vi icon, I think it would be unreadable if you put 4 lines. I think 3 lines are enough. You could use a 13 or 15 inch laptop to verify this.

2. If the connection failed, your comparison in the timeout case might give you "empty and empty", and get the True value (as a misleading). Do you notice this?

3. At the judgement part, the answer is an array with only one element (True). Under this condition, you do not need an array.


[testCase/testTwoServersReportInternalError.vi]

1. For the vi icon, I think it would be unreadable if you put 4 lines. I think 3 lines are enough. You could use a 13 or 15 inch laptop to verify this.

2. Assume you are unlucky, the flat structure of connection side will take (2000 + 5000 * 2 + 1000 + 1000 + 1000) = 15000 ms but your case structure timeouts at 5000 ms.


[testCase/testTwoServersReportLogMessages.vi]

1. For the vi icon, I think it would be unreadable if you put 4 lines. I think 3 lines are enough. You could use a 13 or 15 inch laptop to verify this.

2. If you only send the event for one time, why do you use a string array to hold the data and compare with an array (which has only one element)? Please use a single string variable only.


[testTwoServersSendCmdStatusToClientsAtSameTime.vi]

1. For the vi icon, I think it would be unreadable if you put 4 lines. I think 3 lines are enough. You could use a 13 or 15 inch laptop to verify this.


[testTwoServersSendEventsToClientsAtSameTime.vi]

1. For the vi icon, I think it would be unreadable if you put 4 lines. I think 3 lines are enough. You could use a 13 or 15 inch laptop to verify this.


[testCase/testTwoServersSendTelemetryAtSameTime.vi]

1. For the vi icon, I think it would be unreadable if you put 4 lines. I think 3 lines are enough. You could use a 13 or 15 inch laptop to verify this.


[tests/ServerInstancesTest/serverInstancesTest.vi]

1. Please consider to put this vi to the button of ServerInstancesTest virtual directory of labview project. Thanks!

teweitsai commented:
I would prefer we talk first to see any unclear point in the review before updating the code. Thanks!

pcortes80 commented:
[Overview]

1. For all the test cases, only the error line of server A will tunnell with a flat structure with waiting time = 2000 ms. I think it is better to let the error line of server B to tunnell to that flat structure as well.
Reply: Done.

[tests/ServerInstancesTest/disconnectTcpClientA.vi]

1. This is the same as disconnectTcpClientB.vi. Maybe rename this to disconnectTcpClient.vi and remove disconnectTcpClientB.vi.
Reply: Done.

2. The "label" in block diagram overlaps with the flat structure. Fix this.
Reply: Fixed.

[tests/ServerInstancesTest/disconnectTcpClientB.vi]

1. As comment above, remove this vi.
Reply: Done.

[tests/ServerInstancesTest/readTcpPacketAsString.vi]

1. Please consider to put the timeout=1000 in block diagram to the front panel as an input (+ tunnel). You might want to be able to tune this in the test cases.
Reply: Done.

[tests/ServerInstancesTest/sendCmdFromTcpClientA.vi]

1. The is the same as sendCmdFromTcpClientB.vi. Maybe rename this to sendCmdFromTcpClient.vi and remove sendCmdFromTcpClientB.vi.
Reply: Done.

2. Rename the variable of connectionID_A to connectionID.
Reply: Done.

3. You do not need the shift register of connectionID on the for-loop. The "TCP Write" will not modify that connection ID. You can just use a terminal.
Reply: Done.

4. Please consider to put the timeout=1000 in block diagram to the front panel as an input (+ tunnel). You might want to be able to tune this in the test cases.
Reply: Done.

5. If you need to two instances to run at the same time, you can pre-allocate them.

[tests/ServerInstancesTest/sendCmdFromTcpClientB.vi]

1. As comment above, remove this vi.
Reply: Done

[tests/ServerInstancesTest/sendEventsFromTcpClient.vi]

1. Please consider to put the timeout=1000 in block diagram to the front panel as an input (+ tunnel). You might want to be able to tune this in the test cases.
Reply: Done

[testCase/testCreateMultipleTcpIpConnectionsServersAandB.vi]

1. After the disconnection between the client and server, you would need to wait some time to let the OS to recover the system resource. Otherwise, you might only get the initial connection successfully and all the others failed.
Reply: Done.

2. Unless there is a good reason, I would prefer to use the shift register for the error line for the for-loop instead of using the tunnel. In addition, I noticed that your called the "Clear Errors.vi" already.
Reply: Done.

[testCase/testTwoClientsSendCmdAtSameTime.vi]

1. For the vi icon, I think it would be unreadable if you put 4 lines. I think 3 lines are enough. You could use a 13 or 15 inch laptop to verify this.

2. What would happen if you get the empty connection ID (connection fails in timeout) for the command or telemetry server?
Reply: I did not add any error case structure because the function sendCmdFromTcpClient.vi has the error case structure internally already. So, if there is an error connection the commands will not be sent.

[testCase/testTwoClientsSendEventsAtSameTime.vi]

1. For the vi icon, I think it would be unreadable if you put 4 lines. I think 3 lines are enough. You could use a 13 or 15 inch laptop to verify this.
Reply: Fixed.

2. There is a logic problem here. You have 5 sec for the connection timeout but only 2 sec for the event case. Do you see the problem?
Reply: Fixed.

3. Your event supposes to listern to two EvtRead events but you only put have one in the case structure. You should have two instead. In addition, if you do not care about the received "event", you can judge the result by receiveTimes=2 as an integet (use a shift register to hold the receiveTimes). This is easier to compare than the string array.
Reply: Done.

4. You need to unregister the events after the event case structure is done and destroy the events to release the system resource.
Reply: Done

[testCase/testTwoClientsSendTelemetryAtSameTime.vi]

1. For the vi icon, I think it would be unreadable if you put 4 lines. I think 3 lines are enough. You could use a 13 or 15 inch laptop to verify this.
Reply: Done.

[testCase/testTwoServersReportConnStatus.vi]

1. For the vi icon, I think it would be unreadable if you put 4 lines. I think 3 lines are enough. You could use a 13 or 15 inch laptop to verify this.
Reply: Done.

2. If the connection failed, your comparison in the timeout case might give you "empty and empty", and get the True value (as a misleading). Do you notice this?
Reply: Done.

3. At the judgement part, the answer is an array with only one element (True). Under this condition, you do not need an array.
Reply: Done.

[testCase/testTwoServersReportInternalError.vi]

1. For the vi icon, I think it would be unreadable if you put 4 lines. I think 3 lines are enough. You could use a 13 or 15 inch laptop to verify this.
Reply: Done.

2. Assume you are unlucky, the flat structure of connection side will take (2000 + 5000 * 2 + 1000 + 1000 + 1000) = 15000 ms but your case structure timeouts at 5000 ms.
Reply: Done.

[testCase/testTwoServersReportLogMessages.vi]

1. For the vi icon, I think it would be unreadable if you put 4 lines. I think 3 lines are enough. You could use a 13 or 15 inch laptop to verify this.
Reply: Done.

2. If you only send the event for one time, why do you use a string array to hold the data and compare with an array (which has only one element)? Please use a single string variable only.
Reply: Done.

[testTwoServersSendCmdStatusToClientsAtSameTime.vi]

1. For the vi icon, I think it would be unreadable if you put 4 lines. I think 3 lines are enough. You could use a 13 or 15 inch laptop to verify this.
Reply: Done.

[testTwoServersSendEventsToClientsAtSameTime.vi]

1. For the vi icon, I think it would be unreadable if you put 4 lines. I think 3 lines are enough. You could use a 13 or 15 inch laptop to verify this.
Reply: Done.

[testCase/testTwoServersSendTelemetryAtSameTime.vi]

1. For the vi icon, I think it would be unreadable if you put 4 lines. I think 3 lines are enough. You could use a 13 or 15 inch laptop to verify this.
Reply: Done.

[tests/ServerInstancesTest/serverInstancesTest.vi]

1. Please consider to put this vi to the button of ServerInstancesTest virtual directory of labview project. Thanks!
Reply: Done.
